### PR TITLE
nsolid: support Heap Sampling

### PR DESCRIPTION
**Pull request by juanarbol**
This patch adds support to the "sampled" heapsnapshot.

This new API methods in N|Solid allow users to sample allocations in the V8 heap; internally it does the calls to the V8 Profiler for start sampling and stop sampling.

This patch requires a new mutex and a new thread map to manage which thread is currently sampling, same as CPU profiler and HeapSnapshot.

### CODE REVIEW

File: src/nsolid.h [hunk 1]
```diff
@@ -1015,6 +1015,33 @@ class NODE_EXTERN Snapshot {
...
+                           Data&&... data);
+
+  /**
+   * @brief  Stops the HeapProfiler allocation sampler
+   * @param envinst SharedEnvInst of thread to take the snapshot from.
+   * @return NSOLID_E_SUCCESS in case of success or a different NSOLID_E_
+   * error value otherwise.
+   */
+  static int StopSampling(SharedEnvInst envinst);
+
```
Comment (santigimeno) [inline src/nsolid.h:1076]: As with the others, we need a sync version of stop

File: src/nsolid.h [hunk 2]
```diff
@@ -1015,6 +1015,33 @@ class NODE_EXTERN Snapshot {
...
+   * @param data variable number of arguments to be propagated to the callback.
+   * @return NSOLID_E_SUCCESS in case of success or a different NSOLID_E_
+   * error value otherwise.
+   */
+  template <typename Cb, typename... Data>
+  static int StartSampling(SharedEnvInst envinst,
+                           uint64_t duration,
+                           Cb&& cb,
+                           Data&&... data);
+
```
Comment (santigimeno) [inline src/nsolid.h]: How do I configure the other parameters allowed for sampling. I propose smthg like this:
```c++
template <typename Cb, typename... Data>
  static int StartSampling(SharedEnvInst envinst,
                           uint64_t duration,
                           Cb&& cb,
                           Data&&... data);
  template <typename Cb, typename... Data>
  static int StartSampling(SharedEnvInst envinst,
                           uint64_t sample_interval,
                           int stack_depth,
                           v8::HeapProfiler::SamplingFlags flags,
                           uint64_t duration,
                           Cb&& cb,
                           Data&&... data);
```
where the former calls with the default values to the latter

Comment (santigimeno) [reply src/nsolid.h]: Also, does redacting the sampling data apply here?

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 3]
```diff
@@ -350,5 +402,169 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
...
+  nsuv::ns_mutex::scoped_lock lock(&snapshotter->in_progress_heap_sampling_);
+  auto it = snapshotter->threads_running_heap_sampling_.find(thread_id);
+  ASSERT(it != snapshotter->threads_running_heap_sampling_.end());
+
+  HeapSamplerStor& stor = it->second;
+  if (sampled == nullptr) {
+    QueueCallback(snapshot_cb,
+                  thread_id,
+                  // TODO(juan) - this is not a heap snapshot failure
+                  heap_profiler::HEAP_SNAPSHOT_FAILURE,
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc:654]: Should we address this issue before merging?

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 4]
```diff
@@ -350,5 +402,169 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
...
+                  std::string());
+  } else {
+    nlohmann::ordered_json profile_json =
+      build_v8_allocation_profile(isolate, sampled);
+    nlohmann::ordered_json samples_json;
+    printf("%s", profile_json.dump().c_str());
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]: debugging left over I guess.

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 5]
```diff
@@ -350,5 +402,169 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
...
+    SharedEnvInst envinst,
+    NSolidHeapSnapshot* snapshotter) {
+  uint64_t thread_id = envinst->thread_id();
+  nsuv::ns_mutex::scoped_lock lock(&snapshotter->in_progress_heap_sampling_);
+  auto it = snapshotter->threads_running_heap_sampling_.find(thread_id);
+  if (it == snapshotter->threads_running_heap_sampling_.end()) {
+    // The profiler was stopped before the timer was triggered
+    return;
+  }
+
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]: We can hit the same issue here that we had with heap profiles: by the time this function is called the original sampling operation might be over and a new one might be in progress. We need unique ids.

**Review by santigimeno — CHANGES_REQUESTED** [review]

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 5 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc]: True that, fixed!

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 4 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc]: Yeap; it is gone now

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 3 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc:654]: This is just a naming convention that I don't considerate appropriate. That's it.

File: src/nsolid.h (hunk 2 above)
Comment (juanarbol) [reply src/nsolid.h]: > Also, does redacting the sampling data apply here?

Nope, this does not include any sensitive information.

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 6]
```diff
@@ -350,5 +539,133 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
...
+    stor.cb(0, profile_json.dump().c_str(), stor.data.get());
+    // Send the empty string to signal the end of the profile
+    stor.cb(0, std::string(), stor.data.get());
+    // Stop the profiler
+    profiler->StopSamplingHeapProfiler();
+  }
+  stor.flags |= kFlagIsDone;
+  snapshotter->threads_running_heap_sampling_.erase(it);
+}
+
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]: We need to delete `sampled` in this code path too.

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 7]
```diff
@@ -350,5 +539,133 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
...
+  } else {
+    nlohmann::ordered_json profile_json =
+      build_v8_allocation_profile(isolate, sampled);
+    nlohmann::ordered_json samples_json;
+    // Send the whole profile to the callback
+    stor.cb(0, profile_json.dump().c_str(), stor.data.get());
+    // Send the empty string to signal the end of the profile
+    stor.cb(0, std::string(), stor.data.get());
+    // Stop the profiler
+    profiler->StopSamplingHeapProfiler();
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]: On this Sync version it's ok to call the cb from JS thread but not in the async one. Please. pass the profile to the N|Solid thread using QueueCallback(), in the same way you do above in the error path.

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 8]
```diff
@@ -350,5 +539,133 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
...
+
+  if (er) {
+    // In case the the thread is already gone, the cpu profile will be stopped
+    // on RemoveEnv, so do nothing here.
+  }
+}
+
+void NSolidHeapSnapshot::stop_sampling_profiler(
+    SharedEnvInst envinst,
+    uint64_t snaphot_id,
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]:
```suggestion
    uint64_t snapshot_id,
```

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 7 above)
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]: The `profiler->StopSamplingHeapProfiler();` should be moved after the else, right? as it has to be done on every code path

**Review by santigimeno — CHANGES_REQUESTED** [review]
This looks much better. Left some more comments

File: src/nsolid.cc [hunk 9]
```diff
@@ -521,6 +521,38 @@ int CpuProfiler::get_cpu_profile_(SharedEnvInst envinst,
                                                    deleter);
 }
 
+int Snapshot::start_allocation_sampling_(SharedEnvInst envinst,
+                                         uint64_t sampleInterval,
+                                         int stackDepth,
+                                         v8::HeapProfiler::SamplingFlags flags,
```
Comment (santigimeno) [inline src/nsolid.cc]: Use snake-case in c++
```suggestion
                                         uint64_t sample_interval,
                                         int stack_depth,
                                         v8::HeapProfiler::SamplingFlags flags,
```

File: src/nsolid.h [hunk 10]
```diff
@@ -1023,6 +1084,15 @@ class NODE_EXTERN Snapshot {
                                           internal::user_data data,
                                           snapshot_proxy_sig proxy);
 
+  static int start_allocation_sampling_(SharedEnvInst envinst,
+                                        uint64_t sampleInterval,
+                                        int stackDepth,
```
Comment (santigimeno) [inline src/nsolid.h]: snake case here too

File: src/nsolid.h [hunk 11]
```diff
@@ -4,6 +4,7 @@
 #include "node.h"
 #include "uv.h"
 
+#include <v8-profiler.h>
 #include <memory>
 #include <string>
```
Comment (santigimeno) [inline src/nsolid.h]: Nit: be consistent

```suggestion
#include "node.h"
#include "uv.h"
#include "v8-profiler.h"

#include <memory>
#include <string>
```

**Review by santigimeno — CHANGES_REQUESTED** [review]

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 6 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc]: Done!

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 7 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc]: Done

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 8 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc]: Oopsie, fixed

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 7 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc]: Yes, fixed

File: src/nsolid.h [hunk 12]
```diff
@@ -1015,14 +1016,83 @@ class NODE_EXTERN Snapshot {
    */
   static int StopTrackingHeapObjectsSync(SharedEnvInst envinst);
 
+  /**
+   * @brief Will start sampling heap allocations.
+   *
+   * @param envinst SharedEnvInst of thread to take the snapshot from.
+   * @param duration duration in milliseconds of the heap profiler after which
+   * the heap snapshot will be returned in the callback.
```
Comment (santigimeno) [inline src/nsolid.h]: I'm not sure this text is correct (looks like the one for trackingHeapObjects)

File: src/nsolid.h [hunk 13]
```diff
@@ -1015,14 +1016,83 @@ class NODE_EXTERN Snapshot {
...
+                           Cb&& cb,
+                           Data&&... data);
+
+  /**
+   * @brief Will start sampling heap allocations.
+   *
+   * @param envinst SharedEnvInst of thread to take the snapshot from.
+   * @param sample_interval
+   * @param stack_depth
+   * @param flags
```
Comment (santigimeno) [inline src/nsolid.h]: We need a description of these fields

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 14]
```diff
@@ -8,11 +8,199 @@ namespace nsolid {
...
+  // Make sure there is a snapshot running
+  if (it == threads_running_heap_sampling_.end())
+    return UV_ENOENT;
+
+  HeapSamplerStor& stor = it->second;
+  // If this condition is reached. This was called by EnvList::RemoveEnv.
+  // It wants to stop any pending snapshot w/ tracking heap object.
+  if (!(stor.flags & kFlagIsSamplingHeap) || stor.flags & kFlagIsDone) {
+    // If no pending trackers, just do nothing
+    // There are not peding snapshots with trackers
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]:
```suggestion
    // There are not pending snapshots with trackers
```

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 15]
```diff
@@ -8,11 +8,199 @@ namespace nsolid {
...
+  // Delete the snapshot from the map
+  threads_running_heap_sampling_.erase(it);
+  return 0;
 }
 
 NSolidHeapSnapshot::~NSolidHeapSnapshot() {
   nsuv::ns_mutex::scoped_lock lock(&in_progress_heap_snapshots_);
   threads_running_snapshots_.clear();
+  nsuv::ns_mutex::scoped_lock lock2(&in_progress_heap_sampling_);
+  threads_running_heap_sampling_.clear();
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]:
```suggestion
  {
    nsuv::ns_mutex::scoped_lock lock(&in_progress_heap_snapshots_);
    threads_running_snapshots_.clear();
  }
  {
    nsuv::ns_mutex::scoped_lock lock2(&in_progress_heap_sampling_);
    threads_running_heap_sampling_.clear();
  }
```

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 16]
```diff
@@ -350,5 +538,142 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
...
+  ASSERT(stor.flags & kFlagIsSamplingHeap);
+
+  // Give control back to the V8 thread
+  int er = RunCommand(envinst,
+                      CommandType::Interrupt,
+                      take_sampled_snapshot,
+                      snapshotter);
+  if (er) {
+    // In case the the thread is already gone, the cpu profile will be stopped
+    // on RemoveEnv, so do nothing here.
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]: Fix the comment to refer to the heap sampling

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 17]
```diff
@@ -350,5 +538,142 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
...
+  uint64_t thread_id = envinst->thread_id();
+  nsuv::ns_mutex::scoped_lock lock(&snapshotter->in_progress_heap_sampling_);
+  auto it = snapshotter->threads_running_heap_sampling_.find(thread_id);
+  if (it == snapshotter->threads_running_heap_sampling_.end()) {
+    // The profiler was stopped before the timer was triggered
+    return;
+  }
+
+  HeapSamplerStor& stor = it->second;
+  if (stor.snapshot_id !=  snapshot_id) {
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc]: Is the linter happy with this?
```suggestion
  if (stor.snapshot_id != snapshot_id) {
```

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 18]
```diff
@@ -350,5 +538,142 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
...
+    QueueCallback(sampling_cb,
+                  thread_id,
+                  0,
+                  profile_json.dump(),
+                  snapshotter);
+    QueueCallback(sampling_cb,
+                  thread_id,
+                  0,
+                  std::string(),
+                  snapshotter);
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc:670]: This looks awful to be honest. We need a streaming api for this too. Can you add it to your backlog?

**Review by santigimeno — CHANGES_REQUESTED** [review]
Almost there

File: src/nsolid/nsolid_heap_snapshot.cc (hunk 18 above)
Comment (juanarbol) [reply src/nsolid/nsolid_heap_snapshot.cc:670]: Will do

File: src/nsolid.h [hunk 19]
```diff
@@ -1015,14 +1016,86 @@ class NODE_EXTERN Snapshot {
...
+  /**
+   * @brief Will start sampling heap allocations.
+   *
+   * @param envinst SharedEnvInst of thread to take the snapshot from.
+   * @param sample_interval frequency at which the heap profiler samples the
+   * heap and captures data
+   * @param stack_depth refers to the depth of the call stack that will be
+   * captured during heap profiling
+   * @param flags additional flags or options that can be used when
+   * starting heap profiling
```
Comment (santigimeno) [inline src/nsolid.h]: Nit: name the units used for `sample_interval` (I think they're bytes). Also add a link to v8docs for the flags enum

File: src/nsolid/nsolid_heap_snapshot.cc [hunk 20]
```diff
@@ -350,5 +542,142 @@ void NSolidHeapSnapshot::snapshot_cb(uint64_t thread_id,
...
+
+  HeapSamplerStor& stor = it->second;
+  // If this condition is reached. This was called by EnvList::RemoveEnv.
+  // It wants to stop any pending snapshot w/ tracking heap object.
+  if (!(stor.flags & kFlagIsSamplingHeap) || stor.flags & kFlagIsDone) {
+    // If no pending trackers, just do nothing
+    // There are not peding snapshots with trackers
+    return;
+  }
+
```
Comment (santigimeno) [inline src/nsolid/nsolid_heap_snapshot.cc:647]: Move the `GetAllocationProfile()`, otherwise we might be leaking memory

**Review by santigimeno — APPROVED** [review]
Once those are addressed we should be good. Good job!

File: test/addons/nsolid-heap-sampler/binding.cc [hunk 21]
```diff
@@ -0,0 +1,151 @@
...
+    Integer::New(isolate, status),
+    String::NewFromUtf8(
+      isolate,
+      it->second.profile.c_str(),
+      NewStringType::kNormal).ToLocalChecked()
+  };
+
+  profiles.erase(it);
+  uv_mutex_unlock(&profiles_map_lock);
+  cb->Call(context, Undefined(isolate), 2, argv);
```
Comment (santigimeno) [inline test/addons/nsolid-heap-sampler/binding.cc]: BTW fix this warning here:
```
make[2]: Leaving directory '/home/sgimeno/nodesource/nsolid/test/addons/nsolid-heap-sampler/build'
../binding.cc: In function ‘void sample_cb(node::nsolid::SharedEnvInst, int)’:
../binding.cc:51:11: warning: ignoring return value of ‘v8::MaybeLocal<v8::Value> v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*)’, declared with attribute ‘warn_unused_result’ [-Wunused-result]
   51 |   cb->Call(context, Undefined(isolate), 2, argv);
      |   ~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/sgimeno/nodesource/nsolid/deps/v8/include/v8.h:33,
                 from /home/sgimeno/nodesource/nsolid/src/node.h:73,
                 from ../binding.cc:1:
/home/sgimeno/nodesource/nsolid/deps/v8/include/v8-function.h:55:43: note: declared here
   55 |   V8_WARN_UNUSED_RESULT MaybeLocal<Value> Call(Local<Context> context,
      |                                           ^~~~
```
Also on the other heap profile addon test.

**Review by santigimeno — APPROVED** [review]
Still LGTM with one comment

**Pull request merged**